go/arrow/compute/arithmetic.go

```diff
@@ -0,0 +1,155 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compute
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/apache/arrow/go/v10/arrow"
+	"github.com/apache/arrow/go/v10/arrow/compute/internal/exec"
+	"github.com/apache/arrow/go/v10/arrow/compute/internal/kernels"
+)
+
+type arithmeticFunction struct {
+	ScalarFunction
+}
+
+func (fn *arithmeticFunction) checkDecimals(vals ...arrow.DataType) error {
+	if !hasDecimal(vals...) {
+		return nil
+	}
+
+	if len(vals) != 2 {
+		return nil
+	}
+
+	op := fn.name[:strings.Index(fn.name, "_")]
```

Comment:
Using string comparison for this is a bit weird, no? Can't you have `arithmeticFunction` be some kind of polymorphic type with different implementations of argument promotion for each subtype/implementation/whatever?